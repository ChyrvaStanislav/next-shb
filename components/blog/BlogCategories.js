import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { Typography, Link } from 'v2_main/components/html';
import CategoryLink from './CategoryLink';
import styles from './BlogCategories.module.scss';

const BlogCategories = ({
  categoryList,
  setCategory,
  categorySearch,
  menuTitle,
  setMeta,
  defaultMeta,
  querySearch,
  postTag,
  hasHighlight,
  isBlogPost,
}) => {
  const getCategoryLink = category => {
    const isActive = (postTag ? category?.id === postTag?.toLowerCase() : categorySearch === category?.id);

    return (
      <CategoryLink
        key={category?.id}
        category={category}
        setCategory={setCategory}
        setMeta={setMeta}
        querySearch={querySearch}
        isActive={isBlogPost ? false : isActive}
        style={classNames(styles.filterLink, {
          [styles.active]: hasHighlight && isActive
        })}
      />
    );
  };

  const getAllStoriesLink = () => {
    const activeAll = !categorySearch && !postTag;
    if (activeAll) {
      return (
        <Typography
          variant="p_links"
          className={classNames(styles.filterLink, {
            [styles.active]: hasHighlight && activeAll
          })}
          key="all"
        >
          {menuTitle}
        </Typography>
      );
    }

    const categoryParam = `?query=${querySearch}`;

    return (
      <Link
        to={`/blog${querySearch ? categoryParam : ''}`}
        onClick={() => {
          setCategory('');
          setMeta(defaultMeta);
        }}
        className={styles.filterLink}
        key="all"
        type="link"
      >
        {menuTitle}
      </Link>
    );
  };

  return (
    <aside className={styles.filterContainer}>
      <Typography type="p" variant="h5" className={styles.title}>
        Topics
      </Typography>
      <ul className={styles.categoryList}>
        <li className={styles.categoryItem}>
          {getAllStoriesLink()}
        </li>
        {categoryList.map(category => (
          <li className={styles.categoryItem} key={category.id}>
            {getCategoryLink(category)}
          </li>
        ))}
      </ul>
    </aside>
  );
};

BlogCategories.propTypes = {
  categoryList: PropTypes.arrayOf(PropTypes.any),
  setCategory: PropTypes.func.isRequired,
  categorySearch: PropTypes.string,
  menuTitle: PropTypes.string,
  setMeta: PropTypes.func.isRequired,
  defaultMeta: PropTypes.shape({}),
  querySearch: PropTypes.string,
  postTag: PropTypes.string,
  hasHighlight: PropTypes.bool,
  isBlogPost: PropTypes.bool,
};

BlogCategories.defaultProps = {
  categoryList: [],
  categorySearch: '',
  menuTitle: '',
  defaultMeta: {},
  querySearch: '',
  postTag: '',
  hasHighlight: true,
  isBlogPost: false,
};

export default BlogCategories;
