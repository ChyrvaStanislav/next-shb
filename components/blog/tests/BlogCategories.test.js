import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import BlogCategories from '../BlogCategories';

const categoryList = [
  {
    meta: {
      title: 'Data Analytics Articles | EPAM SolutionsHub',
      description: 'Read useful articles about Data Analytics. '
    },
    label: 'Analytics',
    id: 'analytics'
  },
  {
    meta: {
      title: 'Cloud Devops Blog Posts | EPAM SolutionsHub',
      description: 'Browse the catalog of published posts related to Cloud and DevOps.'
    },
    label: 'Cloud & DevOps',
    id: 'cloud_devops'
  }
];

const defaultMeta = {
  title: 'Blog | EPAM SolutionsHub',
  description: 'Visit our Blog page to read more about Lean Delivery, Testing, Cloud & DevOps, OSCI and other SolutionsHub topics'
};

const setCategory = jest.fn();
const setMeta = jest.fn();

const Component = (props) => (
  <BrowserRouter>
    <BlogCategories
      menuTitle="All Posts"
      categoryList={categoryList}
      setCategory={setCategory}
      setMeta={setMeta}
      defaultMeta={defaultMeta}
      hasHighlight={true}
      {...props}
    />
  </BrowserRouter>
);

describe('BlogCategories snapshots', () => {
  test('should BlogCategories match snapshot with ALL active link', () => {
    const { asFragment } = render(<Component />);

    expect(asFragment()).toMatchSnapshot();
  });

  test('should BlogCategories match snapshot with ONE active link', () => {
    const { asFragment } = render(<Component categorySearch="analytics" postTag="analytics" />);

    expect(asFragment()).toMatchSnapshot();
  });
});

describe('BlogCategories functions', () => {
  test('should BlogCategories onClick by "All posts" link trigger setCategories and seMeta with default data', () => {
    render(<Component />);
    fireEvent.click(screen.getByText('Analytics'));

    expect(setCategory).toHaveBeenCalledWith('analytics');
  });

  test('should BlogCategories onClick by  trigger setMeta', () => {
    render(<Component categorySearch="analytics" postTag="analytics" />);

    fireEvent.click(screen.getByText('All Posts'));

    expect(setCategory).toHaveBeenCalledWith('');
    expect(setMeta).toHaveBeenCalledWith(defaultMeta);
  });
});
