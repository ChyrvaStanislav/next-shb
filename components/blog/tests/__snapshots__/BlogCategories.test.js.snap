// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlogCategories snapshots should BlogCategories match snapshot with ALL active link 1`] = `
<DocumentFragment>
  <aside
    class="filterContainer"
  >
    <p
      class="h5 title"
      other=""
    >
      Topics
    </p>
    <ul
      class="categoryList"
    >
      <li
        class="categoryItem"
      >
        <p
          class="p_links filterLink active"
          other=""
        >
          All Posts
        </p>
      </li>
      <li
        class="categoryItem"
      >
        <a
          class="p_links filterLink link2 topicAnalytics"
          data-testid="category-link"
          href="/blog?category=analytics"
          other=""
        >
          Analytics
        </a>
      </li>
      <li
        class="categoryItem"
      >
        <a
          class="p_links filterLink link2 topicCloud&DevOps"
          data-testid="category-link"
          href="/blog?category=cloud_devops"
          other=""
        >
          Cloud & DevOps
        </a>
      </li>
    </ul>
  </aside>
</DocumentFragment>
`;

exports[`BlogCategories snapshots should BlogCategories match snapshot with ONE active link 1`] = `
<DocumentFragment>
  <aside
    class="filterContainer"
  >
    <p
      class="h5 title"
      other=""
    >
      Topics
    </p>
    <ul
      class="categoryList"
    >
      <li
        class="categoryItem"
      >
        <a
          class="p_links filterLink"
          href="/blog"
          other=""
        >
          All Posts
        </a>
      </li>
      <li
        class="categoryItem"
      >
        <p
          class="p_links filterLink active link2 topicAnalytics"
          other=""
        >
          Analytics
        </p>
      </li>
      <li
        class="categoryItem"
      >
        <a
          class="p_links filterLink link2 topicCloud&DevOps"
          data-testid="category-link"
          href="/blog?category=cloud_devops"
          other=""
        >
          Cloud & DevOps
        </a>
      </li>
    </ul>
  </aside>
</DocumentFragment>
`;
